feat: add Time To Initial Display and Time To Full Display spans (#850)
getsentry/develop: src/docs/sdk/performance/time-to-initial-full-display.mdx

---

> title: "Time to Initial Display/Time to Full Display"

Reviewer: `m`: I like the title form the sidebar more.
```suggestion
title: "Time to Initial/Full Display"
```

---

We recommend implementing this feature for mobile and desktop SDKs.

> The current screen load transactions timings are not reliable to understand the time spent in the screen creation.
> This is because they wait for any span created in the meantime, including any file I/O or network request.

Reviewer: `l`:  A bit cleaner
```suggestion
The current screen load transactions timings are unreliable in understanding the time spent in the screen creation.
This is because they wait for any span created, including any file I/O or network request.
```

Time to Initial Display and Time to Full Display are spans added to the screen load transactions,
which can be used to show more accurate information on the time spent in the screen creation.
Being spans, they don't impact on user quota.

> They start as soon as a new screen is created, as long as the screen load auto-generated transactions are enabled.

Reviewer: `m`: That is a bit unclear to me. I don't fully understand this sentence.

Author: I rephrased it. Is it better now?

## Time to Initial Display

Time to Initial Display is a span with operation `ui.load.initial_display` which finishes automatically
when the screen draws its first frame. It's controlled automatically by the SDK and doesn't
require any manual intervention from the user.

It's also set as a measurement of the transaction, with `time_to_initial_display` as key and the span duration as value.

> Note: The measurement is should be set only if the span finishes with an `OK`  status.

Reviewer: `m`: How could the TTID span end with a different status?

Author: The case is when a new screen starts before showing the first frame
Like an Activity that is used to only start other activities

Reviewer: Please add this essential piece of information to the docs.

## Time to Full Display

Time to Full Display is a span with operation `ui.load.full_display` which is finished manually through
the `Sentry.reportFullyDisplayed` static API. This feature has to be enabled by setting the SDK config

> option `enableTimeToFullDisplayTracing`. Remember that the screen load transactions wait for spans to finish,

Reviewer: `m`: Isn't it called `enableTimeToFullDisplay`?

Author: on Android it's `io.sentry.traces.time-to-full-display.enable` in the manifest and `enableTimeToFullDisplayTracing` in the code, to be consistent with the other options currently available.
Guess we can rephrase it to `enableTimeToFullDisplay` in the docs anyway?

Reviewer: All the other options for controlling other automatic instrumentation have the `tracing` suffix. I think we should name the option `enableTimeToFullDisplayTracing`.

so the transaction will run until the user calls the API. So we recommend this feature to be disabled by default.
It's also set as a measurement of the transaction, with `time_to_full_display` as key and the span duration as value.

> The following section explain the possible scenarios of the Time to Full Display.

Reviewer: `l`:
```suggestion
The following section explains the possible scenarios of the Time to Full Display.
```

The specification is written in the [Gherkin syntax](https://cucumber.io/docs/gherkin/reference/).

```Gherkin
Scenario: Option disabled
    Given an ongoing screen load/navigation transaction

>     Then the Full Display span is not started

Reviewer: `m`: Again please use the active voice. I'm going to make only one comment for the passive voice for the scenarios.
```suggestion
    Then the SDK doesn't start the Full Display span
```

    When the user calls the manual API `reportFullyDisplayed`
    Then the SDK ignores the call

Scenario: Option enabled
    Given the enableTimeToFullDisplayTracing option is enabled
    When the SDK creates a new screen load transaction
    Then the Full Display span is started

Scenario: Activity displayed

>     Given an ongoing Full Display span

Reviewer: `l`: A bit more specific
```suggestion
    Given an ongoing Full Display span
    And the SDK finished the  Initial Display span
```

    When the user calls the API `reportFullyDisplayed`
    Then the SDK finishes the span
    And sets the status to ok

Scenario: API not called
    Given an ongoing Full Display span
    When the user doesn't call the API `reportFullyDisplayed` in 30 seconds
    Then the SDK finishes the span

>     And sets the status to deadline_exceeded

Reviewer: `m`:
```suggestion
    And sets the status to deadline_exceeded
    And sets the timestamp to the timestamp of the Initial Display span
```

Scenario: API called too early
    Given an ongoing Full Display span and an ongoing Initial Display span
    When the user calls the API `reportFullyDisplayed`

>     Then the SDK finishes the Full Display span

Reviewer: `m`: What status does the SDK set?

    And
    When the Initial Display span is finished

>     Then the SDK adjusts the Full Display span end timestamp to the Initial Display span end timestamp

Reviewer: `m`: Please also specify which status the SDK sets to both spans

Scenario: Another screen load/navigation transaction occurs
    Given an ongoing Full Display span
    When a new screen load/navigation transaction is started
    Then the SDK finishes the Full Display span

>     And sets the status to deadline_exceeded

Reviewer: `m`: What is the value of the timestamp?

---
title: "Time to Initial/Full Display"
---

We recommend implementing this feature for mobile and desktop SDKs.

The current screen load transactions timings are unreliable in understanding the time spent in the screen creation.
This is because they wait for any span created, including any file I/O or network request.
Time to Initial Display and Time to Full Display are spans added to the screen load transactions,
which can be used to show more accurate information on the time spent in the screen creation.

> Being spans, they don't impact on user quota.

Reviewer: `l`:
```suggestion
Being spans, they don't impact user quota.
```

```Gherkin
Scenario: Option disabled
    Given an ongoing screen load/navigation transaction
    Then the SDK doesn't start the Full Display span
    When the user calls the manual API `reportFullyDisplayed`
    Then the SDK ignores the call

Scenario: Option enabled
    Given the enableTimeToFullDisplay option is enabled
    When the SDK creates a new screen load transaction
    Then the Full Display span is started

Scenario: Screen displayed
    Given an ongoing Full Display span
    And the SDK finished the  Initial Display span
    When the user calls the API `reportFullyDisplayed`
    Then the SDK finishes the span
    And sets the status to ok

Scenario: API not called
    Given an ongoing Full Display span
    When the user doesn't call the API `reportFullyDisplayed` in 30 seconds
    Then the SDK finishes the span
    And sets the status to deadline_exceeded
    And sets the timestamp to the timestamp of the Initial Display span

Scenario: API called too early
    Given an ongoing Full Display span and an ongoing Initial Display span
    When the user calls the API `reportFullyDisplayed`
    Then the SDK finishes the Full Display span

>     And set the status to ok

Reviewer: `l`:
```suggestion
    And sets the status to ok
```

    And

>     When the Initial Display span is finished

Reviewer: `m`: You should only have one `When` per scenario. You split this scenario up into two.

```Gherkin
Scenario: Option disabled
    Given an ongoing screen load/navigation transaction
    Then the SDK doesn't start the Full Display span
    When the user calls the manual API `reportFullyDisplayed`
    Then the SDK ignores the call

Scenario: Option enabled
    Given the enableTimeToFullDisplay option is enabled
    When the SDK creates a new screen load transaction
    Then the Full Display span is started

Scenario: Screen displayed
    Given an ongoing Full Display span
    And the SDK finished the  Initial Display span
    When the user calls the API `reportFullyDisplayed`
    Then the SDK finishes the span
    And sets the status to ok

Scenario: API not called
    Given an ongoing Full Display span
    When the user doesn't call the API `reportFullyDisplayed` in 30 seconds
    Then the SDK finishes the span
    And sets the status to deadline_exceeded
    And sets the timestamp to the timestamp of the Initial Display span

Scenario: API called
    Given an ongoing Full Display span
    When the user calls the API `reportFullyDisplayed`
    Then the SDK finishes the Full Display span
    And set the status to ok

Scenario: API called too early

>     Given an ongoing Initial Display span and the user calls the API `reportFullyDisplayed`
>     When the Initial Display span is finished

Reviewer: ```suggestion
    Given an ongoing Initial Display span
    And the user called the API `reportFullyDisplayed`
    When the SDK finishes the Initial Display span
```
